[V1][Metrics] Add initial Prometheus logger

markmc opened the pull request:
Part of #10582

Implement the `vllm:num_requests_running` and `vllm:num_requests_waiting` gauges from V0. This is a simple starting point from which to iterate towards parity with V0.

There's no need to use prometheus_client's "multi-processing mode" (at least at this stage) because these metrics all exist within the API server process.

Note this restores the following metrics - these were lost when we started using multi-processing mode:

- python_gc_objects_collected_total
- python_gc_objects_uncollectable_total
- python_gc_collections_total
- python_info
- process_virtual_memory_bytes
- process_resident_memory_bytes
- process_start_time_seconds
- process_cpu_seconds_total
- process_open_fds
- process_max_fds

References:
- #1890 - Add Production Metrics in Prometheus format
- #2316 - Refactor Prometheus and Add Request Level Metrics (v0)
- #2730 - Port metrics from aioprometheus to prometheus_client (v0)
- #7279 - Fix Prometheus Metrics With zeromq Frontend (v0)
- #11962 - Logging and Metrics (v1)
- #11973 - Logging and Metrics - OutputProcessor Abstraction (v1)


With this change:

```
$ python3 ./benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-8B-Instruct --tokenizer meta-llama/Llama-3.1-8B-Instruct --endpoint /v1/completions --dataset-name sharegpt --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --request-rate 3.0 --num-prompts 200
...
============ Serving Benchmark Result ============
Successful requests:                     200
Benchmark duration (s):                  69.09
Total input tokens:                      42659
Total generated tokens:                  43287
Request throughput (req/s):              2.89
Output token throughput (tok/s):         626.49
Total Token throughput (tok/s):          1243.90
---------------Time to First Token----------------
Mean TTFT (ms):                          24.03
Median TTFT (ms):                        20.63
P99 TTFT (ms):                           51.97
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          9.22
Median TPOT (ms):                        9.18
P99 TPOT (ms):                           10.48
---------------Inter-token Latency----------------
Mean ITL (ms):                           9.16
Median ITL (ms):                         8.95
P99 ITL (ms):                            18.07
==================================================
```

With this change and --disable-log-stats

```
$ python3 ./benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-8B-Instruct --tokenizer meta-llama/Llama-3.1-8B-Instruct --endpoint /v1/completions --dataset-name sharegpt --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --request-rate 3.0 --num-prompts 200
...
============ Serving Benchmark Result ============
Successful requests:                     200
Benchmark duration (s):                  69.09
Total input tokens:                      42659
Total generated tokens:                  43463
Request throughput (req/s):              2.89
Output token throughput (tok/s):         629.07
Total Token throughput (tok/s):          1246.50
---------------Time to First Token----------------
Mean TTFT (ms):                          24.77
Median TTFT (ms):                        21.55
P99 TTFT (ms):                           53.24
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          9.23
Median TPOT (ms):                        9.19
P99 TPOT (ms):                           10.44
---------------Inter-token Latency----------------
Mean ITL (ms):                           9.17
Median ITL (ms):                         8.97
P99 ITL (ms):                            18.26
==================================================
```

markmc committed:
[V1][Metrics] Add simple prometheus logger

Part of #10582

Implement the vllm:num_requests_running and vllm:num_requests_waiting
gauges from V0. This is a simple starting point from which to iterate
towards parity with V0.

There's no need to use prometheus_client's "multi-processing mode"
(at least at this stage) because these metrics all exist within the
API server process.

Note this restores the following metrics - these were lost when we
started using multi-processing mode:

- python_gc_objects_collected_total
- python_gc_objects_uncollectable_total
- python_gc_collections_total
- python_info
- process_virtual_memory_bytes
- process_resident_memory_bytes
- process_start_time_seconds
- process_cpu_seconds_total
- process_open_fds
- process_max_fds

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Add v1 test to tests/entrypoints/openai/test_metrics.py

Not possible to use the run_with_both_engines fixture with its monkey-patching
approach, since these tests use a module-level server fixture and the
monkey-patching fixture cannot be used with a module-level fixture. Instead
we can just pass env_dict = {'VLLM_USE_V1': '1'} to RemoteOpenAIServer, but
copy the general approach of run_with_both_engines otherwise.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

joerunde commented on tests/entrypoints/openai/test_metrics.py:
```diff
@@ -84,7 +104,9 @@ async def client(server):
-                              client: openai.AsyncClient):
+                              client: openai.AsyncClient, use_v1: bool):
+    if use_v1:
+        pytest.skip("Skipping test on vllm V1")
```
🤔 Won't this spend the time to build the `server` fixture before just skipping the test?

We could use the `run_with_both_engines` fixture so that `@skip_v1` can be used here, and do 
```
for metric in (EXPECTED_METRICS_V1 if os.getenv('VLLM_USE_V1') == "1" else EXPECTED_METRICS):
```
in the next test where we need to split the logic for v1. It's not nearly as clean but that would let us skip the tests during collection instead of at runtime after fixture setup

markmc replied:
Thanks for the review, Joe

The `server` fixture is a "module-level" fixture - it gets initialized once and shared across the three tests

So the v1 server will always be used by one test, even though the other two gets skipped

And because `server` is a module-level fixture, we can not use function-level fixtures like `run_with_both_engines` or `monkeypatch`

Nor can a function-level decorator cause a module-level decorator to skip a specific test function

I went the `run_with_both_engines` route initially and fumbled my way into the above understanding of the constraints

joerunde replied:
Oh yeah you're right, I missed the server definition in this file 👍

markmc commented:
Looks like a legit test failure, will investigate:

```
ValueError: Duplicated timeseries in CollectorRegistry: {'vllm:num_requests_running'}
```

robertgshaw2-redhat commented:
Hey @markmc, this looks good.

I forget the details of what exactly this does, but we have this in V0 for the CI/CD:
- https://github.com/vllm-project/vllm/blob/324960a95c00112ce6b9b858d9311da1597cfb8b/vllm/engine/metrics.py#L50

markmc committed:
[V1][Metrics] Fix "Duplicated timeseries" error in test_async_llm.py

Fixes:

ValueError: Duplicated timeseries in CollectorRegistry: {'vllm:num_requests_running'}

Same solution as in v0 - in case there are multiple engine instances
in the same process (only in tests?), just de-register the metrics
before registering them.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
TPU Test:

> ERROR: No matching distribution found for torch==2.6.0.dev20241126+cpu

DarkLight1337 commented:
Should be fixed by https://github.com/vllm-project/vllm/pull/12422, can you merge from main?

markmc commented:
> Should be fixed by #12422, can you merge from main?

Thanks!

mgoin reviewed (approved):
Manually tested by following the example: https://github.com/vllm-project/vllm/blob/103bd17ac585b44372a47f365d80f13446cf362d/examples/online_serving/prometheus_grafana/README.md

Checked `http://localhost:8000/metrics` to see `num_requests_running` updating during benchmarking
